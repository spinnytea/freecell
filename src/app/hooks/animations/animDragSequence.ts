import { MutableRefObject } from 'react';
import { gsap } from 'gsap/all';
import { BOTTOM_OF_CASCADE } from '@/app/components/cards/constants';
import { calcCardId } from '@/app/game/card/card';

// FIXME keep refining tale
export function animDragSequence({
	timeline,
	// FIXME we need first card coords, not the position of the pointer/cursor/mouse
	pointerCoords: { x, y },
	list,
	offsetTop,
	gameBoardIdRef,
}: {
	timeline: gsap.core.Timeline;
	list: string[];
	pointerCoords: { x: number; y: number };
	offsetTop: number;
	gameBoardIdRef?: MutableRefObject<string>;
}) {
	const positions = list.map((shorthand) => {
		const el = document.getElementById(calcCardId(shorthand, gameBoardIdRef?.current));
		if (el) {
			const bounds = el.getBoundingClientRect();
			return {
				top: bounds.top,
				left: bounds.left,
			};
		}
		return {
			top: y,
			left: x,
		};
	});

	list.forEach((shorthand, index) => {
		const cardId = '#' + calcCardId(shorthand, gameBoardIdRef?.current);
		const zIndex = BOTTOM_OF_CASCADE + index;
		gsap.set(cardId, { zIndex });
		if (index > 0) {
			// follow the card this is stacked on top of
			// an attempt at follow-the-leader style drag animation, each card lags behind the previous
			const previousCardPosition = positions[index - 1];
			const top = previousCardPosition.top + offsetTop;
			const left = previousCardPosition.left;
			const duration = 2 + index / 10; // REVIEW (drag-and-drop) timing
			timeline.to(cardId, { top, left, duration, ease: 'linear' }, '<0');
		}
	});
}
