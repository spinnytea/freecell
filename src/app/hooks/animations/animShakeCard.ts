import { MutableRefObject } from 'react';
import { gsap } from 'gsap/all';
import { INVALID_SHAKE_MAGNITUDE, INVALID_SHAKE_PORTION } from '@/app/animation_constants';

export function animShakeCard({
	timeline,
	list,
	gameBoardIdRef,
}: {
	timeline: gsap.core.Timeline;
	list: string[];
	gameBoardIdRef?: MutableRefObject<string>;
}) {
	// sometimes we start left, sometimes we start right
	// yoyo is opposite
	const start: boolean = gsap.utils.random([true, false]);
	const signStart = start ? '-' : '+';
	const signYoyo = start ? '+' : '-';

	list.forEach((shorthand, index) => {
		const cardId = '#c' + shorthand + (gameBoardIdRef?.current ? '-' + gameBoardIdRef.current : '');
		const tl = gsap.timeline();
		// offset left
		tl.to(cardId, {
			x: `${signStart}=${INVALID_SHAKE_MAGNITUDE.toFixed(3)}`,
			duration: 0.1,
			ease: 'sine.in',
		});
		// yoyo
		tl.to(cardId, {
			x: `${signYoyo}=${(INVALID_SHAKE_MAGNITUDE * 2).toFixed(3)}`,
			duration: INVALID_SHAKE_PORTION,
			ease: 'sine.inOut',
			yoyo: true,
			repeat: 2,
		});
		// back to center
		tl.to(cardId, { x: '0', duration: INVALID_SHAKE_PORTION / 2, ease: 'sine.out' });

		timeline.add(tl, index === 0 ? `>0` : `<+${INVALID_SHAKE_PORTION.toFixed(3)}`);
	});
}
