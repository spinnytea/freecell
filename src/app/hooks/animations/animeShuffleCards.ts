import { MutableRefObject } from 'react';
import { gsap } from 'gsap/all';
import { SHUFFLE_DURATION, SHUFFLE_R, SHUFFLE_X, SHUFFLE_Y } from '@/app/animation_constants';
import { BOTTOM_OF_CASCADE } from '@/app/components/cards/constants';
import { calcCardId, Card, shorthandCard } from '@/app/game/card/card';

// TODO (techdebt) (animation) review shuffle animation
export function animShuffleCards({
	timeline,
	list,
	gameBoardIdRef,
}: {
	timeline: gsap.core.Timeline;
	list: Card[];
	gameBoardIdRef?: MutableRefObject<string>;
}) {
	let shRight: string | undefined = undefined;
	let shLeft = undefined;
	if (list.length === 1) {
		shRight = shorthandCard(list[0]);
	} else if (list.length === 2) {
		shLeft = shorthandCard(list[0]);
		shRight = shorthandCard(list[1]);
	} else if (list.length > 2) {
		shLeft = shorthandCard(list[1]);
		shRight = shorthandCard(list[2]);
	}

	if (shLeft) {
		const cardId = '#' + calcCardId(shLeft, gameBoardIdRef?.current);
		const tl = gsap.timeline();

		tl.set(cardId, { zIndex: -BOTTOM_OF_CASCADE - 1 });

		tl.to(cardId, {
			x: `-=${SHUFFLE_X.toFixed(3)}`,
			y: `-=${SHUFFLE_Y.toFixed(3)}`,
			rotation: -SHUFFLE_R,
			duration: SHUFFLE_DURATION / 2,
			ease: 'sine.out',
		});
		tl.to(cardId, {
			x: '0',
			y: '0',
			rotation: 0,
			duration: SHUFFLE_DURATION / 2,
			ease: 'sine.in',
		});

		timeline.add(tl, 'shuffle');
	}

	if (shRight) {
		const cardId = '#' + calcCardId(shRight, gameBoardIdRef?.current);
		const tl = gsap.timeline();

		tl.set(cardId, { zIndex: -BOTTOM_OF_CASCADE });

		tl.to(cardId, {
			x: `+=${SHUFFLE_X.toFixed(3)}`,
			y: `-=${SHUFFLE_Y.toFixed(3)}`,
			rotation: SHUFFLE_R,
			duration: SHUFFLE_DURATION / 2,
			ease: 'sine.in',
		});
		tl.to(cardId, {
			x: '0',
			y: '0',
			rotation: 0,
			duration: SHUFFLE_DURATION / 2,
			ease: 'sine.out',
		});

		timeline.add(tl, 'shuffle');
	}
}
