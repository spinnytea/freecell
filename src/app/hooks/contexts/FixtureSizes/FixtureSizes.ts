import {
	BOTTOM_OF_CASCADE,
	CARD_FACE_CUTOFF,
	scale_height,
} from '@/app/components/cards/constants';
import { CardLocation, CardSequence, getRankForCompare, Rank } from '@/app/game/card/card';

// IDEA (hud) bad layout idea: free cells on left (top down), foundation on right (top down)
//  - so tableau can start at the top of the screen?
//  - it's a layout
export const DEFAULT_CLIENT_WIDTH = 800;
export const DEFAULT_CLIENT_HEIGHT = 600;
export type FixtureLayout = 'wide' | 'justified' | 'auto';

/*
	Spacing around cards is all in percentages.
	The cards will by scaled to fit the screen, so too must the negative space.

	LR spacing is expressed in % of card width
	TB spacing is expressed in % of card height
	i.e. TB_CASCADE_OFFSET will always "just barely show the rank" because the svg is also scaled and of fixed layout
*/
const _LR_HOME_MARGIN = 1;
const _LR_HOME_GAP = 2;
const TB_HOME_TOP = 0.2;
const _LR_HOME_CARD_SPACING = 1 / 6;
const TB_TABLEAU_TOP = 0.3;
const _LR_TABLEAU_CARD_SPACING = 2 / 7;
const TB_CASCADE_OFFSET_STANDARD = 0.2;
const TB_CASCADE_OFFSET_SMOL = 0.3;
export const PEEK_UP = 0.25;
export const PEEK_DOWN = 0.5;

export interface FixtureSizes {
	boardWidth: number;
	boardHeight: number;
	cardWidth: number;
	cardHeight: number;

	home: {
		top: number;
		cellLeft: number[];
		foundationLeft: number[];
	};

	tableau: {
		top: number;
		cascadeLeft: number[];
		offsetTop: number;
	};

	deck: {
		top: number;
		left: number;
	};
}

function times<T>(count: number, cb: (i: number) => T): T[] {
	const list: T[] = [];
	for (let i = 0; i < count; i++) list.push(cb(i));
	return list;
}

export function calcFixtureSizes({
	boardWidth,
	boardHeight,
	cellCount = 4,
	cascadeCount = 8,
	fixtureLayout = 'auto',
}: {
	boardWidth?: number;
	boardHeight?: number;
	cellCount?: number;
	cascadeCount?: number;
	fixtureLayout?: FixtureLayout;
}): FixtureSizes {
	if (!boardWidth) boardWidth = DEFAULT_CLIENT_WIDTH;
	if (!boardHeight) boardHeight = DEFAULT_CLIENT_HEIGHT;
	const foundationCount = 4;
	const homeCount = foundationCount + cellCount;
	const homeGapCount = foundationCount - 1 + cellCount - 1;
	const cascadeGapCount = cascadeCount - 1;

	// scale LR_HOME_MARGIN based on aspect ratio
	// wider playing fields need a larger margin
	// taller playing fields can have a smaller margin
	let aspectratio = boardWidth / boardHeight;

	// wide layouts try to ensure we can see the card stacks, so we need to balanace the aspect ratio
	// it's ideally targetting 16:9 or 21:9, so we can just stick with that
	if (fixtureLayout === 'wide') aspectratio = 2;
	// justified layouts will fit the width, regardless of height
	else if (fixtureLayout === 'justified') aspectratio = 0.001;
	// smooth transition across the +/-, but negative (effectively) doubles it's scale factor
	else if (aspectratio < 1) aspectratio = 1 - (1 - aspectratio) * 2.5;

	// less spacing in portait mode, more spacing in landscape
	const LR_HOME_CARD_SPACING_SCALED =
		_LR_HOME_CARD_SPACING * Math.max(0.5, Math.min(aspectratio, 2));

	// min margins of the original card spacing
	// otherwise scale up or down
	const LR_HOME_MARGIN_SCALED = Math.max(_LR_HOME_CARD_SPACING, _LR_HOME_MARGIN * aspectratio);

	// leave at least one space for the deck (…but don't actually put it there, the deal looks lame)
	// leave at most 2 card widths
	// otherwise, scale down the gap with the aspect ratio
	const LR_HOME_GAP_SCALED = Math.max(
		1 + LR_HOME_CARD_SPACING_SCALED * 2,
		Math.min(_LR_HOME_GAP * aspectratio, 2 + LR_HOME_CARD_SPACING_SCALED * 2)
	);

	// cannot be wider than the home row
	// use default spacing, or scale down to be as wide as
	const LR_TABLEAU_CARD_SPACING_SCALED = Math.min(
		_LR_TABLEAU_CARD_SPACING,
		(LR_HOME_CARD_SPACING_SCALED * homeGapCount + LR_HOME_GAP_SCALED) / cascadeGapCount
	);

	// cells gap foundation
	// this takes up the most space, so it determines the width of the cards
	//   lr + 4 cards (3 spaces) + gap + 4 cards (3 spaces) + lr = boardWidth
	//   spaces                                                       +  cards               = boardWidth
	// ((LR_HOME_MARGIN * 2 + LR_HOME_GAP + LR_HOME_CARD_SPACING_SCALED * 6) + (1 * 8)) * cardWidth = boardWidth
	const cardWidth =
		boardWidth /
		(LR_HOME_MARGIN_SCALED * 2 +
			LR_HOME_GAP_SCALED +
			LR_HOME_CARD_SPACING_SCALED * homeGapCount +
			homeCount);

	const cardHeight = scale_height(cardWidth);

	const TB_CASCADE_OFFSET =
		cardWidth < CARD_FACE_CUTOFF ? TB_CASCADE_OFFSET_SMOL : TB_CASCADE_OFFSET_STANDARD;

	// lr + 8 cards (7 spaces) + lr = boardWidth
	// adjust tableau margins to center cards (in px values)
	const tableauLRMargin =
		(boardWidth - (cascadeCount + LR_TABLEAU_CARD_SPACING_SCALED * cascadeGapCount) * cardWidth) /
		2;

	const cellLOffset = LR_HOME_MARGIN_SCALED * cardWidth;
	const foundationLOffset =
		(LR_HOME_MARGIN_SCALED +
			LR_HOME_CARD_SPACING_SCALED * (cellCount - 1) +
			cellCount +
			LR_HOME_GAP_SCALED) *
		cardWidth;

	return {
		boardWidth: toFixed(boardWidth),
		boardHeight: toFixed(boardHeight),
		cardWidth: toFixed(cardWidth),
		cardHeight: toFixed(cardHeight),

		home: {
			top: toFixed(TB_HOME_TOP * cardHeight),
			// left justified
			// < lr + cell + space + cell + …
			cellLeft: times(cellCount, (i) =>
				toFixed((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth * i + cellLOffset)
			),
			// left justified
			// gap << cell + space + cell + …
			foundationLeft: times(foundationCount, (i) =>
				toFixed((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth * i + foundationLOffset)
			),

			// right justified
			// … + cell + space + cell + lr >
			// foundationLeft: [
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth) * 3 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth) * 2 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth) * 1 -
			// 		cellLOffset),
			// 	toFixed(boardWidth -
			// 		cardWidth -
			// 		((1 + LR_HOME_CARD_SPACING_SCALED) * cardWidth) * 0 -
			// 		cellLOffset),
			// ],
		},

		tableau: {
			top: toFixed((TB_HOME_TOP + 1 + TB_TABLEAU_TOP) * cardHeight),
			// left justified
			// < lr + cell + space + cell + …
			cascadeLeft: times(cascadeCount, (i) =>
				toFixed((1 + LR_TABLEAU_CARD_SPACING_SCALED) * cardWidth * i + tableauLRMargin)
			),
			offsetTop: toFixed(TB_CASCADE_OFFSET * cardHeight),
		},

		deck: {
			top: toFixed(boardHeight - cardHeight - TB_HOME_TOP * cardHeight),
			left: toFixed(cellLOffset),
		},
	};
}

function toFixed(num: number): number {
	return parseFloat(num.toFixed(3));
}

export function calcTopLeftZ(
	fixtureSizes: FixtureSizes,
	{ fixture, data }: CardLocation,
	selection: CardSequence | null,
	rank?: Rank
): { top: number; left: number; zIndex: number; rotation: number } {
	switch (fixture) {
		case 'deck':
			// TODO (animation) animate cursor (selection?) within deck
			return {
				top: fixtureSizes.deck.top,
				left: fixtureSizes.deck.left,
				zIndex: data[0],
				rotation: 0,
			};
		case 'cell':
			return {
				top: fixtureSizes.home.top,
				left: fixtureSizes.home.cellLeft[data[0]],
				zIndex: data[0],
				rotation:
					selection?.location.fixture === 'cell' && selection.location.data[0] === data[0] ? 10 : 0,
			};
		case 'foundation':
			return {
				top: fixtureSizes.home.top,
				left: fixtureSizes.home.foundationLeft[data[0]],
				// zIndex + BOTTOM_OF_CASCADE so it's above the cell/cascade
				// REVIEW (animation) cards in flight
				zIndex: (rank ? getRankForCompare(rank) : 0) + BOTTOM_OF_CASCADE,
				rotation:
					selection?.location.fixture === 'foundation' && selection.location.data[0] === data[0]
						? 10
						: 0,
			};
		case 'cascade': {
			const ret = {
				top: fixtureSizes.tableau.top + fixtureSizes.tableau.offsetTop * data[1],
				left: fixtureSizes.tableau.cascadeLeft[data[0]],
				zIndex: data[1], // we don't really need to make one cascade strictly above another
				rotation: 0,
			};
			if (selection?.location.fixture === 'cascade' && selection.location.data[0] === data[0]) {
				const sd1 = selection.location.data[1];
				if (selection.cards.length > 1 || !selection.canMove) {
					if (data[1] > sd1) {
						ret.top += fixtureSizes.tableau.offsetTop * PEEK_DOWN;
					} else if (data[1] === sd1 && sd1 > 0) {
						ret.top -= fixtureSizes.tableau.offsetTop * PEEK_UP;
					}
				} else if (data[1] === sd1) {
					ret.top += fixtureSizes.tableau.offsetTop * PEEK_DOWN;
				}
			}
			return ret;
		}
	}
}
