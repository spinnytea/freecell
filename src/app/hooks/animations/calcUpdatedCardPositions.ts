import {
	Card,
	CardSequence,
	Fixture,
	getRankForCompare,
	shorthandCard,
	Suit,
} from '@/app/game/card/card';
import { parsePreviousActionMoveShorthands } from '@/app/game/move/history';
import { calcTopLeftZ, FixtureSizes } from '@/app/hooks/contexts/FixtureSizes/FixtureSizes';

export interface UpdateCardPositionsType {
	shorthand: string;
	top: number;
	left: number;
	zIndex: number;
	rank: number;
	suit: Suit;
	previousTop: number;
}

// TODO (techdebt) (combine-move-auto-foundation) unit test
// TODO (animation) (motivation) optimize
export function calcUpdatedCardPositions({
	fixtureSizes,
	previousTLs,
	cards,
	selection,
	actionText,
	actionPrev,
}: {
	fixtureSizes: FixtureSizes;
	previousTLs: Map<string, number[]>;
	cards: Card[];
	selection: CardSequence | null;
	actionText?: string;
	actionPrev?: Card[];
}): {
	updateCardPositions: UpdateCardPositionsType[];
	updateCardPositionsPrev?: UpdateCardPositionsType[];
	secondMustComeAfter?: boolean;
} {
	const updateCardPositions: UpdateCardPositionsType[] = [];
	const fixtures = new Set<Fixture>();

	cards.forEach((card) => {
		const { top, left, zIndex } = calcTopLeftZ(fixtureSizes, card.location, selection, card.rank);
		const shorthand = shorthandCard(card);

		const prev = previousTLs.get(shorthand);
		if (!prev || prev[0] !== top || prev[1] !== left) {
			updateCardPositions.push({
				shorthand,
				top,
				left,
				zIndex,
				rank: getRankForCompare(card.rank),
				suit: card.suit,
				previousTop: prev?.[0] ?? top,
			});
			fixtures.add(card.location.fixture);
		}
	});

	if (!updateCardPositions.length || !actionText) {
		return { updateCardPositions };
	}

	// IFF all of the cards moving are the same as the ones in action text (all of a in b, all of b in a);
	//  - then move a to old spot, then move ALL to new spot
	if (actionPrev) {
		// len(update) = len(union(move, auto)) -> winning may auto what we just moved, so move+auto > update
		//  - could include some or all of move (e.g. if you move a sequence and only part gets auto)
		//  - if we undo we could have more or less cards
		//    we need to make sure all the cards in question are in the list
		//    AND we need to make sure our two lists cover everything in updateCardPositions
		const { moveShorthands, autoFoundationShorthands } =
			parsePreviousActionMoveShorthands(actionText);
		if (
			moveShorthands &&
			updateCardPositions.length <= moveShorthands.length + autoFoundationShorthands.length
		) {
			const { updateCardPositions: prevUpdateCardPositions } = calcUpdatedCardPositions({
				fixtureSizes,
				previousTLs,
				cards: actionPrev,
				selection: null,
			});

			let anyMissing = false;
			const a = moveShorthands.map((sh) => {
				const position = prevUpdateCardPositions.find(({ shorthand }) => shorthand === sh);
				if (!position) anyMissing = true;
				return position;
			});

			// filter items from updateCardPositions if they are in A and have exactly the same position
			let secondMustComeAfter = false;
			const b = updateCardPositions.filter(({ shorthand, top, left }) => {
				const found = a.find((_a) => _a?.shorthand === shorthand);
				if (!found) return true;
				if (found.top !== top) {
					secondMustComeAfter = true;
					return true;
				}
				if (found.left !== left) {
					secondMustComeAfter = true;
					return true;
				}
				return false;
			});

			// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
			if (!anyMissing) {
				return {
					updateCardPositions: b,
					updateCardPositionsPrev: a as UpdateCardPositionsType[],
					secondMustComeAfter,
				};
			}
		}
	}

	// fallback to something simple
	const isFoundation = fixtures.size === 1 && fixtures.has('foundation');
	if (isFoundation) {
		// order by rank / top
		// REVIEW (animation) this needs more work
		//  - can we parse the previous action for the card list?? that's in the correct order!
		//  - that works moving forward, but undo is all crazy
		//  - i guess we can default to "top" if the lists don't match
		// ---
		//  - no matter what tricks we apply, the auto-foundation animation will _always_ be wrong if we do not finish the previous animation first
		//  - animate((g) => g.touch()).animate((g) => g.autoFoundation())
		// REVIEW (animation) dynamic overlap? start of slow and then speed up, / accelerate
		// IDEA (motivation) (animation) different animations for "auto-foundation" vs "win" vs "flourish" (can just check previousAction.type)
		// IDEA (animation) auto-foundation win needs more drama than just "do the same thing"
		// IDEA (animation) flourish: first card goes up. then second card goes up. then third card overlaps abit ... second-to-last AND last go up at the same time
		updateCardPositions
			.sort((a, b) => a.previousTop - b.previousTop)
			.sort((a, b) => a.rank - b.rank);
	} else {
		// order by top
		updateCardPositions.sort(({ top: a }, { top: b }) => a - b);
	}

	return { updateCardPositions };
}
