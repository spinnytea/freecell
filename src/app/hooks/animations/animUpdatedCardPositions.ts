import { MutableRefObject } from 'react';
import { gsap } from 'gsap/all';
import {
	DEFAULT_TRANSLATE_DURATION,
	MAX_ANIMATION_OVERLAP,
	TOTAL_DEFAULT_MOVEMENT_DURATION,
} from '@/app/animation_constants';
import { UpdateCardPositionsType } from '@/app/hooks/animations/calcUpdatedCardPositions';
import { FixtureSizes } from '@/app/hooks/contexts/FixtureSizes/FixtureSizes';

export function animUpdatedCardPositions({
	timeline,
	list,
	nextTLs,
	fixtureSizes,
	prevFixtureSizes,
	gameBoardIdRef,
	pause = false,
}: {
	timeline: gsap.core.Timeline;
	list: UpdateCardPositionsType[];
	nextTLs: Map<string, number[]>;
	fixtureSizes: FixtureSizes;
	prevFixtureSizes: MutableRefObject<FixtureSizes>;
	gameBoardIdRef?: MutableRefObject<string>;
	pause?: boolean;
}) {
	let overlap = Math.min(
		(TOTAL_DEFAULT_MOVEMENT_DURATION - DEFAULT_TRANSLATE_DURATION) / list.length,
		MAX_ANIMATION_OVERLAP
	);
	if (prevFixtureSizes.current !== fixtureSizes) {
		// XXX (techdebt) (animation) should this just do gsap.set ?
		overlap = 0;
		prevFixtureSizes.current = fixtureSizes;
	}
	list.forEach(({ shorthand, top, left, zIndex }, index) => {
		const cardId = '#c' + shorthand + (gameBoardIdRef?.current ? '-' + gameBoardIdRef.current : '');
		const prevTL = nextTLs.get(shorthand);
		nextTLs.set(shorthand, [top, left]);
		if (prevTL) {
			if (!pause) {
				// bugfix: timeline.to should be enough, but mobile sometimes remakes cards at 0,0
				//  - timeline.fromTo ensures we start the animation from the actual previous place
				timeline.fromTo(
					cardId,
					{ top: prevTL[0], left: prevTL[1] },
					{ top, left, duration: DEFAULT_TRANSLATE_DURATION, ease: 'power1.out' },
					index === 0 ? `>0` : `<${overlap.toFixed(3)}`
				);
			} else {
				// bugfix: but if the same card is moving in two animations,
				// we need to wait for the first to finish before we can start the second
				// and the `.fromTo` is screwing with things, so fall back to just a `.to`
				timeline.to(
					cardId,
					{ top, left, duration: DEFAULT_TRANSLATE_DURATION, ease: 'power1.out' },
					index === 0 ? `>${overlap.toFixed(3)}` : `<${overlap.toFixed(3)}`
				);
			}
			// REVIEW (animation) zIndex boost while in flight?
			//  - as soon as it starts moving, set 100 + Math.max(prevZIndex, zIndex)
			//  - as soon as it finishes animating, set it to the correct value
			timeline.to(cardId, { zIndex, duration: DEFAULT_TRANSLATE_DURATION / 2, ease: 'none' }, `<`);
		} else {
			// when we draw the cards for the first time, don't animate them from (0, 0)
			// for gameplay, this should just be drawing the deck
			timeline.set(cardId, { top, left, zIndex });
		}
	});
}
