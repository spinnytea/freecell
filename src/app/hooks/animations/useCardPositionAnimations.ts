import { MutableRefObject, useRef } from 'react';
import { useGSAP } from '@gsap/react';
import { gsap } from 'gsap/all';
import { MULTI_ANIMATION_TIMESCALE } from '@/app/animation_constants';
import { calcCardId } from '@/app/game/card/card';
import { animShakeCard } from '@/app/hooks/animations/animShakeCard';
import { animUpdatedCardPositions } from '@/app/hooks/animations/animUpdatedCardPositions';
import { calcUpdatedCardPositions } from '@/app/hooks/animations/calcUpdatedCardPositions';
import { useFixtureSizes } from '@/app/hooks/contexts/FixtureSizes/useFixtureSizes';
import { useGame } from '@/app/hooks/contexts/Game/useGame';

// IDEA (settings) setting for "reduced motion" - disable most animations
// IDEA (animation) faster "select-to-peek" animation - when the cards are shifting to peek the selected card
export function useCardPositionAnimations(gameBoardIdRef?: MutableRefObject<string>) {
	const { cards, selection, previousAction } = useGame();
	const fixtureSizes = useFixtureSizes();

	/**
		keep track of card positions, we need to animate anything that moves
		if it hasn't moved since last time, then we don't need to animate it

		IDEA (animation) (drag-and-drop) (techdebt) Store previous positions on DOM? Data attr? Context? GSAP (like, hault an animation, and continue from current position)?
		 - It's just t/l
		 - accessor method for unit testing
		---
		 - a local previousTLs doesn't play with other animations
		 - drag-and-drop needs to know card positions too


		IDEA (animation) (techdebt) Initial positions
		 - Animations don't have positions until they change
		 - So it animates everything all at once (fine)
		 - Iff we don't have previousTLs, then g.undo() and seed it with those.

		BUG (animation) (techdebt) Refresh and then immediately "new game" doesn't animate correctly (possibly because there are no saved card positions l yet)
	*/
	const previousTLs = useRef(new Map<string, number[]>());

	/**
		if we change the size of the screen, then everything will animate
		don't do any offsets, just move/update all the cards immediately

		positions of cards are controlled entirely by the animates

		TODO (techdebt) change to "usePrevious" which can/should return hasChanged
	*/
	const prevFixtureSizes = useRef(fixtureSizes);

	/**
		handle on the currently running / previous animation
		if we have one running when we start a new one,
		then finish it and kill it immediately
		this is a user-interactive game, and we want to prioritize responsiveness
	*/
	const previousTimeline = useRef<gsap.core.Timeline | null>(null);

	useGSAP(
		() => {
			const timeline = gsap.timeline({
				onComplete: () => {
					previousTimeline.current = null;
				},
			});

			const {
				updateCardPositions,
				updateCardPositionsPrev,
				secondMustComeAfter,
				unmovedCards,
				invalidMoveCards,
			} = calcUpdatedCardPositions({
				fixtureSizes,
				previousTLs: previousTLs.current,
				cards,
				selection,
				previousAction,
			});

			if (updateCardPositions.length) {
				if (previousTimeline.current && previousTimeline.current !== timeline) {
					// FIXME test
					previousTimeline.current.timeScale(MULTI_ANIMATION_TIMESCALE); // speed up the previous animations
					// previousTimeline.current
					// 	.totalProgress(1) // jump to the end of the animation (no tweening, no timing, just get there)
					// 	.kill(); // stop animating
				}
				previousTimeline.current = timeline;

				const nextTLs = new Map(previousTLs.current);
				unmovedCards.forEach(({ shorthand, top, left, zIndex }) => {
					// XXX (animation) should this be in animUpdatedCardPositions somehow?
					const cardId = '#' + calcCardId(shorthand, gameBoardIdRef?.current);
					timeline.set(cardId, { top, left, zIndex });
				});
				if (updateCardPositionsPrev) {
					// XXX (techdebt) (motivation) this needs to be refactored this is the first non-trivial animation, so it's a bit of a 1-off
					//  - everything else so far has been about making sure the cards move in the right order
					timeline.addLabel('updateCardPositionsPrev');
					animUpdatedCardPositions({
						timeline,
						list: updateCardPositionsPrev,
						nextTLs,
						fixtureSizes,
						prevFixtureSizes,
						gameBoardIdRef,
					});
				}
				timeline.addLabel('updateCardPositions');
				animUpdatedCardPositions({
					timeline,
					list: updateCardPositions,
					nextTLs,
					fixtureSizes,
					prevFixtureSizes,
					gameBoardIdRef,
					pause: secondMustComeAfter,
				});
				previousTLs.current = nextTLs;
			}

			if (invalidMoveCards?.fromShorthands.length) {
				if (previousTimeline.current && previousTimeline.current !== timeline) {
					// FIXME test
					previousTimeline.current.timeScale(MULTI_ANIMATION_TIMESCALE);
					// previousTimeline.current
					// 	.totalProgress(1) // jump to the end of the animation (no tweening, no timing, just get there)
					// 	.kill(); // stop animating
				}
				previousTimeline.current = timeline;

				timeline.addLabel('invalidMoveCards.fromShorthands');
				animShakeCard({
					timeline,
					list: invalidMoveCards.fromShorthands,
					gameBoardIdRef,
				});
				if (invalidMoveCards.toShorthands.length) {
					timeline.addLabel('invalidMoveCards.toShorthands');
					animShakeCard({
						timeline,
						list: invalidMoveCards.toShorthands,
						gameBoardIdRef,
					});
				}
			}
		},
		{ dependencies: [cards, selection, previousAction, fixtureSizes] }
	);

	return { previousTLs };
}
