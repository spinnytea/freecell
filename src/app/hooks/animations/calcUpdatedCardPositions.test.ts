import { Card } from '@/app/game/card/card';
import { calcUpdatedCardPositions } from '@/app/hooks/animations/calcUpdatedCardPositions';
import { calcFixtureSizes } from '@/app/hooks/contexts/FixtureSizes/FixtureSizes';

const fixtureSizes = calcFixtureSizes({
	cellCount: 4,
	cascadeCount: 8,
	fixtureLayout: 'portrait',
});

// FIXME test.todo
describe('animations/calcUpdatedCardPositions', () => {
	// there are a lot of variables, we just need to _start_
	test('first', () => {
		const previousTLs = new Map<string, number[]>([['KH', [-123, 0]]]);
		const cards: Card[] = [
			{
				rank: 'king',
				suit: 'hearts',
				location: {
					fixture: 'cell',
					data: [1],
				},
			},
		];

		const { updateCardPositions, updateCardPositionsPrev, secondMustComeAfter } =
			calcUpdatedCardPositions({
				fixtureSizes,
				previousTLs,
				cards,
				selection: null,
				actionText: 'init',
			});

		expect(updateCardPositions).toMatchSnapshot();
		expect(updateCardPositions.map(({ shorthand }) => shorthand)).toEqual(['KH']);
		expect(updateCardPositionsPrev).toBe(undefined);
		expect(secondMustComeAfter).toBe(undefined);
	});

	test.todo('everything');
});
