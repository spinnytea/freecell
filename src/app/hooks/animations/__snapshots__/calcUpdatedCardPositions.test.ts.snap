// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animations/calcUpdatedCardPositions first 1`] = `
[
  {
    "left": 105.263,
    "previousTop": -123,
    "rank": 12,
    "shorthand": "KH",
    "suit": "hearts",
    "top": 24.4,
    "zIndex": 1,
  },
]
`;
